sdk/azidentity/cache/linux.go

```diff
@@ -0,0 +1,242 @@
+			if err == nil {
+				log.Write(azidentity.EventAuthentication, msg+". Falling back to unencrypted storage")
+				return file.New(f)
+			}
+		} else {
+			err = errors.New(msg + ". Set AllowUnencryptedStorage on TokenCachePersistenceOptions to store the cache in plaintext instead of returning this error")
+		}
+		return nil, err
+	}
+	return newKeyring(name)
+}
+
+// keyring encrypts cache data with a key stored on the user keyring and writes the encrypted
+// data to a file. The encryption key, and thus the data, is lost when the system shuts down.
+type keyring struct {
+	description, file string
+	key               []byte
+	keyID, ringID     int
+}
+
+func newKeyring(name string) (*keyring, error) {
+	p, err := internal.CacheFilePath(name)
+	if err != nil {
+		return nil, err
+	}
+	// the user keyring is available to all processes owned by the user whereas the user
+	// *session* keyring is available only to processes in the current session i.e. shell
+	ringID, err := unix.KeyctlGetKeyringID(unix.KEY_SPEC_USER_KEYRING, true)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't get the user keyring due to error %q", err)
+	}
+	// Attempt to link a persistent keyring to the user's plain old keyring. This enables adding
+	// "persistent" keys that survive all the user's login sessions being deleted. Like all other
+	// keys, persistent keys exist only in kernel memory and are therefore lost on shutdown. If
+	// the attempt fails--some systems don't support persistent keyrings--we just use the plain
+	// old keyring.
+	if persistentRing, err := unix.KeyctlInt(unix.KEYCTL_GET_PERSISTENT, -1, ringID, 0, 0); err == nil {
+		ringID = persistentRing
+	}
+	return &keyring{description: name, file: p, ringID: ringID}, nil
+}
+
+func (k *keyring) Delete(context.Context) error {
+	if k.keyID != 0 && k.ringID != 0 {
+		_, err := unix.KeyctlInt(unix.KEYCTL_UNLINK, k.keyID, k.ringID, 0, 0)
+		if err != nil && !isKeyInvalidOrNotFound(err) {
+			return fmt.Errorf("failed to delete cache data due to error %q", err)
+		}
+	}
+	err := os.Remove(k.file)
+	if errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+	return err
+}
+
+func (k *keyring) Read(context.Context) ([]byte, error) {
+	b, err := os.ReadFile(k.file)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	if len(b) == 0 {
+		return nil, nil
+	}
+	j, err := jwe.ParseCompactFormat(b)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't parse cache data due to error %q", err)
+	}
+	plaintext, err := k.decrypt(j)
+	return plaintext, err
+}
+
+func (k *keyring) Write(_ context.Context, data []byte) error {
+	if len(data) == 0 {
+		return nil
+	}
+	j, err := k.encrypt(data)
+	if err != nil {
+		return err
+	}
+	content, err := j.Serialize()
+	if err != nil {
+		return err
+	}
+	err = os.WriteFile(k.file, []byte(content), 0600)
+	if errors.Is(err, os.ErrNotExist) {
+		err = os.MkdirAll(filepath.Dir(k.file), 0700)
```

Comment:
You could call this before `os.WriteFile()`, it's a no-op if the dir exists.

Reply:
It doesn't change the filesystem in that case but it still makes a syscall, which I imagine is more expensive than `errors.Is` (haven't measured it though)